### PR TITLE
Send requests and events using outbox

### PR DESCRIPTION
**Pull request by mgrachev**

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -93,5 +95,20 @@ pub struct MetricsConfig {
 
 #[derive(Clone, Debug, Deserialize)]
 pub struct MetricsHttpConfig {
-    pub bind_address: std::net::SocketAddr,
+    pub bind_address: SocketAddr,
+}
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+pub struct OutboxConfig {
```
Comment (manifest) [inline src/config.rs:102]: This struct is a part of the `outbox` module.

File: src/config.rs [hunk 2]
```diff
@@ -93,5 +95,20 @@ pub struct MetricsConfig {
...
+    pub messages_per_try: i64,
+    #[serde(with = "crate::serde::duration_seconds")]
+    pub try_wake_interval: chrono::Duration,
+    #[serde(with = "crate::serde::duration_seconds")]
+    pub max_delivery_interval: chrono::Duration,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct NatsConfig {
```
Comment (manifest) [inline src/config.rs:111]: This struct is a part of the `notification-client` crate.

File: src/outbox/pipeline/diesel.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            });
+
+            match result {
+                Ok(Some(next_id)) => {
+                    id = next_id;
+                }
+                Ok(None) => {
+                    break;
+                }
+                Err(err) => return Err(err),
```
Comment (manifest) [inline src/outbox/pipeline/diesel.rs:178]: Why don't we use `?` operator?

File: src/outbox/pipeline/mod.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            f,
+            "pipeline error, kind: {:?}, error: {:?}",
+            self.kind, self.error
+        )
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[allow(clippy::enum_variant_names)]
+pub enum Error {
```
Comment (manifest) [inline src/outbox/pipeline/mod.rs:56]: Consider placing `Error`, `PipelineError`, and `StateError` into a single module `outbox/error`.

_Placing `StateError` to the `outbox` module and `Error` to the `outbox/pipeline` module is misleading._

File: src/outbox/pipeline/diesel.rs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        loop {
+            let conn = self.get_conn().await?;
+
+            let result = conn.transaction(|| {
+                let records: Vec<(Object, T)> =
+                    Self::load_multiple_records(&conn, records_per_try)?;
+
+                if records.is_empty() {
+                    // Exit from the closure
+                    return Ok::<_, Error>(None);
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/outbox/pipeline/diesel.rs:200]: Yes, otherwise we get the compilation error:
```
error[E0283]: type annotations needed
   --> src/outbox/pipeline/diesel.rs:201:28
    |
194 |             let result = conn.transaction(|| {
    |                               ----------- type must be known at this point
...
201 |                     return Ok(None);
    |                            ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
    |
    = note: cannot satisfy `_: From<diesel::result::Error>`
note: required by a bound in `transaction`
   --> /Users/mgrachev/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-1.4.8/src/connection/mod.rs:98:12
    |
98  |         E: From<Error>,
    |            ^^^^^^^^^^^ required by this bound in `transaction`
help: consider specifying the type arguments in the function call
    |
194 |             let result = conn.transaction::<T, E, F>(|| {
    |                                          +++++++++++

For more information about this error, try `rustc --explain E0283`.
error: could not compile `conference` due to previous error
```

File: src/outbox/pipeline/diesel.rs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                for (record, stage) in records {
+                    // In case of error, we try to handle another record
+                    let _ = self.handle_record(&conn, ctx.clone(), record, stage);
+                }
+
+                Ok(Some(()))
+            })?;
+
+            // Break the loop if there are no records
+            if result.is_none() {
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/outbox/pipeline/diesel.rs:212]: Yes, we can't break the loop from the closure `conn.transaction(|| {})`.

File: src/outbox/pipeline/diesel.rs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                    Self::load_multiple_records(&conn, records_per_try)?;
+
+                if records.is_empty() {
+                    // Exit from the closure
+                    return Ok::<_, Error>(None);
+                }
+
+                for (record, stage) in records {
+                    // In case of error, we try to handle another record
+                    let _ = self.handle_record(&conn, ctx.clone(), record, stage);
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/outbox/pipeline/diesel.rs]: Now we log errors in the [run_multiple_stages](https://github.com/foxford/conference/pull/407/commits/a1803be1e1a11d26929eb1fc07a4056b2531b4a2#diff-b74dccec7304a18a98db6ced93286baa174f847884a1f4a126878dbb539b6f71R205) function.

File: src/outbox/pipeline/diesel.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/outbox/pipeline/diesel.rs]: [Updated](https://github.com/foxford/conference/pull/407/commits/a1803be1e1a11d26929eb1fc07a4056b2531b4a2#diff-b74dccec7304a18a98db6ced93286baa174f847884a1f4a126878dbb539b6f71R199).

File: src/outbox/pipeline/diesel.rs (hunk 3 above)
Comment (mgrachev) [reply src/outbox/pipeline/diesel.rs]: [Fixed](https://github.com/foxford/conference/pull/407/commits/a1803be1e1a11d26929eb1fc07a4056b2531b4a2#diff-b74dccec7304a18a98db6ced93286baa174f847884a1f4a126878dbb539b6f71R169).

File: src/config.rs (hunk 1 above)
Comment (mgrachev) [reply src/config.rs]: [Done](https://github.com/foxford/conference/pull/407/commits/d77e8a091894cfcff23a1492a98751ff828483a8#diff-5d8a5ec4280114debab77f8763322f2906751d10206c939bbcf68587472a4a5fR4).

File: src/config.rs (hunk 2 above)
Comment (mgrachev) [reply src/config.rs]: [Done](https://github.com/foxford/svc-nats-client/pull/1/commits/36c7ca7464bb0ee1902903c810e6f0c28d5c1a6f#diff-cba64c21ab992eaad29fce147a08f4560a4769bc14682b8a96081a5fd02dbecdR4).

File: src/outbox/pipeline/mod.rs (hunk 4 above)
Comment (mgrachev) [reply src/outbox/pipeline/mod.rs]: [Done](https://github.com/foxford/conference/pull/407/commits/25f945ffd6a6c6c429cf41496727acd21a157a00#diff-4603f53f83296a6da1ad811d34e041a705c68d0c0ee24b4b3ce2f5988efe3abe).

File: src/app/endpoint/mod.rs [hunk 8]
```diff
@@ -59,7 +59,6 @@ request_routes!(
     "message.unicast" => message::UnicastHandler,
     "room.close" => room::CloseHandler,
     "room.create" => room::CreateHandler,
-    "room.enter" => room::EnterHandler,
```
Comment (0nkery) [inline src/app/endpoint/mod.rs:62]: Why is this removed?

File: src/app/endpoint/room.rs [hunk 9]
```diff
@@ -2001,7 +2041,7 @@ mod test {
             assert_eq!(group_agent.groups().len(), 1);
         }
 
-        #[tokio::test]
+        #[tokio::test(flavor = "multi_thread")]
```
Comment (0nkery) [inline src/app/endpoint/room.rs:2122]: Why it should be multi-threaded?

File: src/app/stage/video_group/send_mqtt_notification.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    type Context = Arc<dyn GlobalContext>;
+    type Stage = AppStage;
+
+    async fn handle(
+        &self,
+        ctx: &Self::Context,
+        _id: &EventId,
+    ) -> Result<Option<Self::Stage>, StageError> {
+        let mqtt_client = ctx.mqtt_client();
+        let mut client = mqtt_client.lock().map_err(|err| {
```
Comment (0nkery) [inline src/app/stage/video_group/send_mqtt_notification.rs:34]: We could use `Mutex` from `parking_lot` since it doesn't have poisoning and is generally faster than std's `Mutex`. It would simplify code.

File: src/db/room.rs [hunk 11]
```diff
@@ -188,25 +188,32 @@ pub trait FindQueryable {
...
-            .filter(room::id.eq(self.id))
-            .get_result(conn)
-            .optional()
+        let mut query = room::table.into_boxed();
+
+        if let Some(id) = self.id {
+            query = query.filter(room::id.eq(id))
+        }
+
+        query.get_result(conn).optional()
```
Comment (0nkery) [inline src/db/room.rs:216]: What this query returns when `id` is `None`?

File: chart/templates/app-cm.yaml [hunk 12]
```diff
@@ -28,6 +28,15 @@ data:
     keep_alive = 15
     max_message_size = 5000000
 
+    [outbox]
+    messages_per_try = 20
+    try_wake_interval = 60 # 1 minute
+    max_delivery_interval = 86400 # 1 day
```
Comment (0nkery) [inline chart/templates/app-cm.yaml:34]: Should it be so long?

File: src/outbox/pipeline/diesel.rs [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    where
+        T: StageHandle<Context = C, Stage = T>,
+        T: Clone + Serialize + DeserializeOwned,
+        C: Send + Sync + 'static,
+    {
+        let id = record.id();
+
+        let rt = tokio::runtime::Handle::current();
+        let event_id = id.clone();
+        let result = std::thread::spawn(move || {
```
Comment (0nkery) [inline src/outbox/pipeline/diesel.rs:97]: Looks too complex.

File: src/outbox/pipeline/diesel.rs [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+        T: StageHandle<Context = C, Stage = T>,
+        T: Clone + Serialize + DeserializeOwned,
+        C: Clone + Send + Sync + 'static,
+    {
+        let mut id = id;
+
+        loop {
+            let conn = self.get_conn().await?;
+
+            let result = conn.transaction(|| {
```
Comment (0nkery) [inline src/outbox/pipeline/diesel.rs:175]: This is blocking while

```rust
  let conn = self.get_conn().await?;
```

is made to be async. Is it correct to call `transaction` in the async context? Maybe we should do something like this:

```rust
spawn_blocking(|| {
  let conn = self.get_conn()?;
  conn.transaction(|| ...);?
});
```

File: src/outbox/pipeline/diesel.rs [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+    ) -> Result<(), PipelineErrors>
+    where
+        T: StageHandle<Context = C, Stage = T>,
+        T: Clone + Serialize + DeserializeOwned,
+        C: Clone + Send + Sync + 'static,
+    {
+        let mut errors = PipelineErrors::new();
+
+        loop {
+            let conn = self.get_conn().await?;
```
Comment (0nkery) [inline src/outbox/pipeline/diesel.rs:207]: Same question as above.

File: src/app/stage/video_group/send_mqtt_notification.rs (hunk 10 above)
Comment (mgrachev) [reply src/app/stage/video_group/send_mqtt_notification.rs]: Agree, [done](https://github.com/foxford/conference/pull/407/commits/e1e61145aca9f7ceb146fd064a0c5bb17d34fce1).

File: src/outbox/error.rs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+pub enum ErrorKind {
+    DbConnAcquisitionFailed,
+    DbQueryFailed,
+    LoadStagesFailed,
+    SerializationFailed,
+    DeserializationFailed,
+    DeleteStageFailed,
+    UpdateStageFailed,
+    InsertStageFailed,
+    StageError(String),
```
Comment (manifest) [inline src/outbox/error.rs:38]: Why the inner type is `String`?

File: src/outbox/error.rs [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+impl PipelineErrors {
+    pub fn new() -> Self {
+        PipelineErrors(vec![])
+    }
+
+    pub fn add(&mut self, error: PipelineError) {
+        self.0.push(error);
+    }
+
+    pub fn is_not_empty(&self) -> bool {
```
Comment (manifest) [inline src/outbox/error.rs:84]: This method name is inconvenient.

Comment (mgrachev) [reply src/outbox/error.rs]: Replaced by `is_empty`.

File: src/app/context.rs [hunk 18]
```diff
@@ -26,7 +29,7 @@ use super::metrics::Metrics;
 
 pub trait Context: GlobalContext + MessageContext {}
 
-pub trait GlobalContext: Sync {
+pub trait GlobalContext: Send + Sync {
```
Comment (manifest) [inline src/app/context.rs:32]: Why do we use any constraint of the trait?

File: src/app/context.rs [hunk 19]
```diff
@@ -69,18 +120,25 @@ pub struct AppContext {
     metrics: Arc<Metrics>,
     mqtt_gateway_client: MqttGatewayHttpClient,
     conference_client: ConferenceHttpClient,
+    mqtt_client: Arc<Mutex<dyn MqttClient>>,
```
Comment (manifest) [inline src/app/context.rs:123]: The best practice would be to make the client struct of a reference type. See `request::Client` for example.

```rust
pub struct Client {
    inner: Arc<ClientRef>,
}
```

This way, the boilerplate goes away and most of the code gets simpler

```rust
mqtt_client: MqttClient
```

File: src/app/mod.rs [hunk 20]
```diff
@@ -142,10 +159,13 @@ pub async fn run(
                 .into_make_service(),
             )
             .with_graceful_shutdown(async move {
-                let _ = graceful_rx.await;
+                shutdown_server_rx.changed().await.ok();
             }),
     );
 
+    let ctx: Arc<dyn GlobalContext> = Arc::new(context.clone());
```
Comment (manifest) [inline src/app/mod.rs:166]: Why don't we just clone Context and pass it to the `outbox_handler`?

File: src/app/context.rs (hunk 19 above)
Comment (0nkery) [reply src/app/context.rs:123]: I think it should be a separate task, too much for this PR.

Comment (manifest) [reply src/app/context.rs:123]: MqttClient was introduced in this PR. That's a bit odd to add something just to fix it later.

File: src/app/context.rs (hunk 18 above)
Comment (mgrachev) [reply src/app/context.rs]: [Removed](https://github.com/foxford/conference/pull/407/commits/8841041d399e9559d67bd7378edcd97c8782fdfa)

File: src/app/mod.rs (hunk 20 above)
Comment (mgrachev) [reply src/app/mod.rs]: We pass `Arc<dyn GlobalContext>` to outbox stages.

**Review by 0nkery — APPROVED** [review]

File: src/app/mod.rs (hunk 20 above)
Comment (mgrachev) [reply src/app/mod.rs]: [Updated](https://github.com/foxford/conference/pull/407/commits/59366f0f7acd9fef4ded858cd42a0267828db9ea)
